### PR TITLE
DUPLO-21305 TF: Support aurora-iopt1 storage type for duplocloud_rds_instance

### PR DESCRIPTION
**Pull request by Duplo-nikv**
## Overview

Added support for aurora-iopt1 for aurora-mysql and aurora-postgresql 
## Summary of changes
Added support for aurora-iopt1 for aurora-mysql and aurora-postgresql with version support duplocloud_rds_instance

This PR does the following:

- Added aurora-iopt1 to storage_type validation
- Updated logic in customdiff function to handle the supported version of storage type and throw user friendly error on plan
- tested for aurora psql on backend branch feturw/DUPLO-21306-supprt-aurora-iopt1-storage-type
- tested for aurora mysql on backend branch feturw/DUPLO-21306-supprt-aurora-iopt1-storage-type
## Testing performed

- [ ] Using unit tests
- [ ] Manually, on my local system
- [ ✓] Manually, on a remote test system

## Describe any breaking changes

- ...

### CODE REVIEW

**Comment by zafarabbas** [comment]
Task linked: [DUPLO-21305 TF: Support aurora-iopt1 storage type for duplocloud_rds_instance](https://app.clickup.com/t/8655600/DUPLO-21305)
Task linked: [DUPLO-21306 BE: Support aurora-iopt1 storage type](https://app.clickup.com/t/8655600/DUPLO-21306)

File: docs/resources/rds_instance.md [hunk 1]
```diff
@@ -373,7 +373,7 @@ If you don't know the available engine versions for your RDS instance, you can u
 - `performance_insights` (Block List, Max: 1) Amazon RDS Performance Insights is a database performance tuning and monitoring feature that helps you quickly assess the load on your database, and determine when and where to take action. Perfomance Insights get apply when enable is set to true. (see [below for nested schema](#nestedblock--performance_insights))
 - `skip_final_snapshot` (Boolean) If the final snapshot should be taken. When set to true, the final snapshot will not be taken when the resource is deleted. Defaults to `false`.
 - `snapshot_id` (String) A database snapshot to initialize the RDS instance from, at launch.
-- `storage_type` (String) Valid values: gp2 | gp3 | io1 | standard | aurora. Storage type to be used for RDS instance storage.
+- `storage_type` (String) Valid values: gp2 | gp3 | io1 | standard | aurora | aurora-iopt1. Storage type to be used for RDS instance storage.
```
Comment (pravin-duplocloud) [inline docs/resources/rds_instance.md:376]: - can you add more on storage_type
- Aurora: 
https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/Concepts.Aurora_Fea_Regions_DB-eng.Feature.storage-type.html
- RDS: 
https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_Storage.html

image

File: duplocloud/utils.go [hunk 2]
```diff
@@ -1012,3 +1013,56 @@ func addIfDefined(target interface{}, resourceName string, targetValue interface
 		}
 	}
 }
+func extractVersion(input string) string {
```
Comment (pravin-duplocloud) [inline duplocloud/utils.go:1016]: -  can you move version utility methods into => version_utils.go

**Review by pravin-duplocloud — CHANGES_REQUESTED** [review]

File: duplocloud/resource_duplo_rds_instance.go [hunk 3]
```diff
@@ -180,12 +180,12 @@ func rdsInstanceSchema() map[string]*schema.Schema {
...
 		},
 		"storage_type": {
-			Description: "Valid values: gp2 | gp3 | io1 | standard | aurora. Storage type to be used for RDS instance storage.",
+			Description: "Valid values: gp2 | gp3 | io1 | standard | aurora | aurora-iopt1. Storage type to be used for RDS instance storage.",
 			Type:        schema.TypeString,
 			Optional:    true,
 			Computed:    true,
 			ValidateFunc: validation.StringInSlice(
-				[]string{"gp2", "gp3", "io1", "standard", "aurora"},
+				[]string{"gp2", "gp3", "io1", "standard", "aurora", "aurora-iopt1"},
```
[the remark this answers is not in the document]
Comment (Duplo-nikv) [reply duplocloud/resource_duplo_rds_instance.go:188]: Its validating value of storage type I added aurora-iopt1 as additional storage type

File: duplocloud/utils.go (hunk 2 above)
Comment (Duplo-nikv) [reply duplocloud/utils.go:1016]: moved from utils to version_utils file

File: docs/resources/rds_instance.md (hunk 1 above)
Comment (Duplo-nikv) [reply docs/resources/rds_instance.md:376]: done

**Review by tahir-duplo — DISMISSED** [review]
LGTM

File: duplocloud/version_utils.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
+package duplocloud
```
Comment (pravin-duplocloud) [inline duplocloud/version_utils.go:1]: already exists 
terraform-provider-duplocloud/duplosdk/version_utils.go
can you merge them

**Review by pravin-duplocloud — DISMISSED** [review]

File: duplocloud/resource_duplo_rds_instance.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pravin-duplocloud) [reply duplocloud/resource_duplo_rds_instance.go:197]: this is how we did it in backend
private void ValidateAuroraIopt1()
{
    if (!string.Equals(StorageType, "aurora-iopt1"))
        return;

    var error = $"Invalid Aurora engine version, StorageType 'aurora-iopt1' not supported for '{EngineVersion}'";
    if (Engine == DBEngine.AuroraMySql && EngineVersion?.IndexOf(".mysql_aurora.") > 0)
    {
        //  '5.7.mysql_aurora.2.11.1'
        var vers = EngineVersion.Split(new[] { ".mysql_aurora." }, StringSplitOptions.None);
        var version = vers.Last();
        var verCurrent = Utils.ParseVersion(version);
        // Aurora MySQL version 3.03.1 and higher
        if (verCurrent.Major > 3)
            return;
        if ( verCurrent.Major < 3
            || string.Equals(version, "3.03.0")
            || !Utils.IsCurrentMajorMinorVersionNewerOrEqualTo(version, "3.03.1"))
            throw new InvalidDataException(error);
    }

    if (Engine == DBEngine.AuroraPostgreSql)
    {
        var verCurrent = Utils.ParseVersion(EngineVersion);
        if (verCurrent.Major > 17)
            return;
        // Aurora PostgreSQL versions 16.1 and higher, 15.2 and higher, 14.7 and higher, and 13.10 and higher
        if (verCurrent.Major < 13
            || verCurrent.Major == 13 && !Utils.IsCurrentMajorMinorVersionNewerOrEqualTo(EngineVersion, "13.10")
            || verCurrent.Major == 14 && !Utils.IsCurrentMajorMinorVersionNewerOrEqualTo(EngineVersion, "14.7")
            || verCurrent.Major == 15 && !Utils.IsCurrentMajorMinorVersionNewerOrEqualTo(EngineVersion, "15.2")
            || verCurrent.Major == 16 && !Utils.IsCurrentMajorMinorVersionNewerOrEqualTo(EngineVersion, "16.1"))
            throw new InvalidDataException(error);
    }
}

**Pull request merged**